File: Problems/0700_0799/0783_Minimum_Distance_Between_BST_Nodes/Project_Go/Operate_TreeNode.go
package main

import (
	"fmt"
	"math"
	"strconv"
)

func setTreeNode(flds []string) *TreeNode {
	return setTreeNode2(flds, 0, 0)
}

func setTreeNode2(flds []string, depth int, pos int) *TreeNode {
	if len(flds) == 0 {
		return nil
	}

	curPos := 0
	for i := 0; i < depth; i++ {
		curPos += int(math.Pow(2.0, float64(i)))
	}

	if curPos+pos > len(flds)-1 {
		return nil
	}

	if flds[curPos+pos] == "null" {
		return nil
	}

	node := new(TreeNode)
	val, err := strconv.Atoi(flds[curPos+pos])

	if err != nil {
		fmt.Printf("setTreeNode() Error ... flds[%d] = %s", curPos+pos, flds[curPos+pos])
		return nil
	}

	node.Val = val
	node.Left = setTreeNode2(flds, depth+1, 2*pos)
	node.Right = setTreeNode2(flds, depth+1, 2*pos+1)

	return node
}

var resultStr = []string{}

func outputTreeNode(node *TreeNode) string {
	resultStr = []string{}

	outStr := setResultStr(node, 0)
	resultStr = nil

	return outStr
}

func setResultStr(node *TreeNode, n int) string {
	if node == nil {
		return ""
	}

	if len(resultStr) <= n {
		resultStr = append(resultStr, "("+strconv.Itoa(node.Val)+")")
	} else {
		resultStr[n] = resultStr[n] + ",(" + strconv.Itoa(node.Val) + ")"
	}

	if node.Left != nil {
		setResultStr(node.Left, n+1)
	}
	if node.Right != nil {
		setResultStr(node.Right, n+1)
	}

	outputStr := ""
	for _, s := range resultStr {
		outputStr += s + "\n"
	}

	return outputStr
}

func Tree2str(t *TreeNode) string {
	if t == nil {
		return ""
	}

	resultStr := strconv.Itoa(t.Val)

	if t.Left == nil && t.Right == nil {
		return resultStr
	}

	resultStr += "(" + Tree2str(t.Left) + ")"
	if t.Right != nil {
		resultStr += "(" + Tree2str(t.Right) + ")"
	}

	return resultStr
}

File: Problems/0500_0599/0538_Convert_BST_to_Greater_Tree/Project_CS/Operate_TreeNode.cs
using System;
using System.Collections.Generic;

public class Operate_TreeNode
{
    public TreeNode set_TreeNode(string[] flds)
    {
         return set_TreeNode(flds, 0, 0);
    }

    public TreeNode set_TreeNode(string[] flds, int depth, int pos)
    {
        if (flds.Length == 0)
            return null;

        int cur_pos = 0;
        for (int i = 0; i < depth; ++i)
            cur_pos += (int)Math.Pow(2, i);
        
        if (cur_pos + pos > flds.Length - 1)
            return null;
        
        if (flds[cur_pos + pos] == "null")
            return null;

        if (flds[cur_pos + pos] == "")
            return null;

        try
        {
            TreeNode node = new TreeNode(int.Parse(flds[cur_pos + pos]));
            node.left = set_TreeNode(flds, depth + 1, 2*pos);
            node.right = set_TreeNode(flds, depth + 1, 2*pos + 1);

            return node;
        }
        catch (Exception e)
        {
            Console.WriteLine("\n" +  e.Message + "\n" +
                              "set_TreeNode() Error ... flds[" + (cur_pos + pos).ToString() + "] = " + flds[cur_pos + pos] + "\n");
            Environment.Exit(-1);

            return null;
        }
    }

    List<string> resultStr;

    public string output_TreeNode(TreeNode node)
    {
        resultStr = new List<string>();

        string outStr = set_ResultStr(node, 0);
        resultStr.Clear();

        return outStr;
    }

    private string set_ResultStr(TreeNode node, int n)
    {
        if (node == null)
            return "";
        
        if (resultStr.Count <= n)
            resultStr.Add("(" + node.val.ToString() + ")");
        else
            resultStr[n] += ",(" + node.val.ToString() + ")";

        if (node.left != null)
            set_ResultStr(node.left, n + 1);
        if (node.right != null)
            set_ResultStr(node.right, n + 1);

        string outputStr = "";
        for (int i = 0; i < resultStr.Count; ++i)
        {
            outputStr += resultStr[i] + "\n";
        }

        return outputStr;
    }

    public string Tree2str(TreeNode t)
    {
        if (t == null)
            return "";

        string resultStr = t.val.ToString();

        if (t.left == null && t.right == null)
            return resultStr;

        resultStr += "(" + Tree2str(t.left) + ")";
        if (t.right != null)
            resultStr += "(" + Tree2str(t.right) + ")";

        return resultStr;
    }
}